Devise導入＋新規登録＋ログイン・ログアウト＋Userモデルの単体テスト

55Suape opened the pull request:
#What
ユーザー管理機能の実装
#Why
ユーザー管理機能を実装するため。以下ご確認をお願いします。

会員登録画面で必須項目を適切に入力した場合の表示
https://gyazo.com/9bb30a6f5f598c8fa27235dca36a2857

会員登録画面で入力に問題がある場合のエラー表示
https://gyazo.com/6043e021bbd8209961ad8966e39af3b1

ログイン画面で必須項目を適切に入力した場合の表示
https://gyazo.com/802fbf9363312e81f08583c80df1e8dc

ログイン画面で入力に問題がある場合のエラー表示
https://gyazo.com/dada37a59e720e32f60512e416127888

ログアウトのイメージ
https://gyazo.com/8e29f5c5a71074753b7965005f10b8d0

Userモデルのテスト結果
https://gyazo.com/777ae574ec95d296be2a43dbfcc77bd7

55Suape committed:
Devise導入＋新規登録＋ログイン・ログアウト＋Userモデルの単体テスト

wad-sho commented on app/models/user.rb:
```diff
@@ -0,0 +1,12 @@
+  validates :first_name, presence: true
+  validates :last_name_kana, presence: true
+  validates :first_name_kana, presence: true
+  validates :birth_date, presence: true
```
```
・パスワードに半角英数字混合での入力を要求するバリデーションも追記しましょう！
・ユーザー名のフリガナは姓名共に、全角（カタカナ）での入力についてバリデーションを追記しましょう！
・ユーザー名は姓名共に、全角（漢字・ひらがな・カタカナ）での入力についてバリデーションを追記しましょう！
```
以下が参考となると存じますので、よければご参照くださいませ！

■参考カリキュラム
応用課題図書 Lesson3正規表現 〜Webアプリケーションに正規表現を応用しよう〜

55Suape replied:
ご連絡いただいた３点の改修を完了しました。

wad-sho commented on app/views/devise/registrations/new.html.erb:
```diff
@@ -11,29 +11,29 @@
   <%# インスタンスを渡して、エラー発生時にメッセージが表示されるようにしましょう。%>
-  <%# render 'shared/error_messages', model: f.object %>
+  <%= render 'shared/error_messages', model: f.object %>
   <%# //インスタンスを渡して、エラー発生時にメッセージが表示されるようにしましょう。%>
```
```
<%# 「モデル名」にはUserモデルであれば@userを渡しましょう。「新規登録機能へのパス」は、devise導入後にrails routesを実行してdevise/registrations#createへのパスを確認し、記載してください。 %>
<%# インスタンスを渡して、エラー発生時にメッセージが表示されるようにしましょう。%>
```

上記の様なコメントアウトは、ユーザー管理機能の実装案内のコメントアウトですので削除しておきましょう！

wad-sho commented on app/views/shared/_header.html.erb:
```diff
@@ -15,12 +15,15 @@
+      <li><%= link_to current_user.nickname, "#", class: "user-nickname" %></li>
+      <li><%= link_to 'ログアウト', destroy_user_session_path, method: :delete, class: "logout" %></li>
+    <% else %>
       <%# //deviseを導入できたら、ログインの有無で表示が変わるように分岐しましょう%>
```
<%# deviseを導入できたら、ログインの有無で表示が変わるように分岐しましょう%>
上記も実装案内ですのでコメントアウトは削除しておきましょう！

wad-sho commented on db/migrate/20230215114233_devise_create_users.rb:
```diff
@@ -0,0 +1,50 @@
+    create_table :users do |t|
+      ## Database authenticatable
+      t.string :nickname,           null: false
+      t.string :email,              null: false, unique: true, default: ""
```
`  t.string :email,              null: false, unique: true, default: ""`

 emailに対するunique: true はdeviseを利用した場合初期設定で付与されますので、ロールバックした上で修正しておきましょう！

wad-sho commented on spec/factories/users.rb:
```diff
@@ -0,0 +1,13 @@
+  factory :user do
+    nickname              {Faker::Name.last_name}
+    email                 {Faker::Internet.free_email}
+    password              {Faker::Internet.password(min_length: 6)}
```
password              {Faker::Internet.password(min_length: 6)}

Fakerでのpasswordの生成について記述できていて素晴らしいです！
こちらの記述ですと、稀に英字のみもしくは数値のみの値が生成される場合があります。
英数字をそれぞれ少なくとも1文字ずつ含む様な形に修正しておきましょう！

wad-sho commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,84 @@
+      end
+    end
+  
+    context '新規登録できないとき' do
```
バリデーションの追記に合わせて、異常系テストも項目の追加をお願いいたします！

wad-sho reviewed (commented):
実装お疲れ様です！レビューのご依頼ありがとうございます。
コメントいたしましたので、ご確認頂けますと幸いでございます。

ご修正後再度レビュー依頼のほどよろしくお願いいたします！

※次回レビューご依頼時には**最新のモデル単体テスト実施結果のキャプチャ**もご提出のほどお願いいたします！

55Suape committed:
Userモデル修正（正規表現を用いたバリデーションの設定）

・パスワードには「半角英数字」混合の入力を要求するバリデーションを追記
・ユーザー名は姓名共に、全角（漢字・ひらがな・カタカナ）での入力を要求するバリデーションを追記
・ユーザー名の「フリガナ」は姓名共に、全角（カタカナ）での入力を要求するバリデーションを追記

55Suape committed:
Viewの修正（ユーザー管理機能に関連するコメントアウト削除）

55Suape committed:
deviseのUserテーブル修正

emailに対するunique: true はdeviseを利用した場合初期設定で付与されるため、ロールバックした上でトルツメ。

55Suape committed:
Userモデルの単体テスト修正＋Rubocop

下記バリデーション追記に伴うテスト項目の追加
・パスワードには「半角英数字」混合の入力を要求するバリデーションを追記
・ユーザー名は姓名共に、全角（漢字・ひらがな・カタカナ）での入力を要求するバリデーションを追記
・ユーザー名の「フリガナ」は姓名共に、全角（カタカナ）での入力を要求するバリデーションを追記

55Suape commented:
#What
ユーザー管理機能の実装修正

#Why
ご連絡いただきました各種修正を完了しましたので、ご確認をお願いします。

Userモデルのテスト結果（バリデーション追記に伴う項目追加）
https://gyazo.com/be487706f4bbff066d55cd4b0c274ca8

ShotaOtake commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,108 @@
+        @user.valid?
+        expect(@user.errors.full_messages).to include('Email is invalid')
+      end
+      it 'passwordは英数字をそれぞれ少なくとも1文字ずつ含まないと登録できない' do
```
■テストの要件を追加しましょう！

現状、下記の要件が満たされていない状況でございます。

```

・英字のみのパスワードでは登録できない
・数字のみのパスワードでは登録できない
・全角文字を含むパスワードでは登録できない

※パスワードにつきましては上記それぞれの項目に分けてexampleの生成をお願い致します。

```

上記を全て満たせるようexampleを追加しましょう！

ShotaOtake reviewed (commented):
レビュー依頼を頂きありがとうございます！
追加で修正依頼のコメントをさせて頂きましたので、ご確認の上修正をお願い致します。

完成まであと一歩です！
引き続き実装頑張っていきましょう！

```
※テストコードにて修正箇所のございます際は、お手数ですが、最新の実行結果の画像も改めて添付頂きますようお願い致します。
```

55Suape committed:
Userモデルの単体テスト修正(不足要件の追加)

パスワードの以下項目を、それぞれ分けてexampleに追加
・英字のみのパスワードでは登録できない
・数字のみのパスワードでは登録できない
・全角文字を含むパスワードでは登録できない

55Suape commented:
#What
ユーザー管理機能の実装修正

#Why
テスト要件について修正を完了しましたので、ご確認をお願いします。

Userモデルのテスト結果（テスト要件の追加）
https://gyazo.com/37077fd04023522f420fd6f403634368

ShotaOtake reviewed (commented):
レビュー依頼を頂きありがとうございます！
お送り頂いた内容を確認致しました。

今回【LGTM】とさせて頂きます。
実装お疲れ様でした！

```
※現在のブランチはマージ後、ローカルのmasterブランチに忘れずにプルしましょう！
その後、新しいブランチを作成し次の実装に取り組んでいただけますと幸いです。
```

引き続き実装頑張っていきましょう！
応援しております！
